commons-numbers-complex/src/main/java/org/apache/commons/numbers/complex/ComplexFunctions.java

```diff
@@ -354,6 +665,1057 @@ private static <R> R log(DoubleUnaryOperator log,
+        return sinh(-imaginary, real, (r, i) -> multiplyNegativeI(r, i, action));
+    }
+
+    /**
+     * Returns the
+     * <a href="http://mathworld.wolfram.com/Cosine.html">
+     * cosine</a> of the complex number.
+     *
+     * <p>\[ \cos(z) = \frac{1}{2} \left( e^{iz} + e^{-iz} \right) \]
+     *
+     * <p>This is an even function: \( \cos(z) = \cos(-z) \).
+     * The cosine is an entire function and requires no branch cuts.
+     *
+     * <p>This is implemented using real \( x \) and imaginary \( y \) parts:
+     *
+     * <p>\[ \cos(x + iy) = \cos(x)\cosh(y) - i \sin(x)\sinh(y) \]
+     *
+     * <p>As per the C99 standard this function is computed using the trigonomic identity:
+     *
+     * <p>\[ cos(z) = cosh(iz) \]
+     *
+     * @param real Real part \( a \) of the complex number \( (a +ib \).
+     * @param imaginary Imaginary part \( b \) of the complex number \( (a +ib \).
+     * @param action Consumer for the cosine of the complex number.
+     * @param <R> the return type of the supplied action.
+     * @return the object returned by the supplied action.
+     * @see <a href="http://functions.wolfram.com/ElementaryFunctions/Cos/">Cos</a>
+     */
+    public static <R> R cos(double real, double imaginary, ComplexSink<R> action) {
+        // Define in terms of cosh
+        // cos(z) = cosh(iz)
+        // Multiply this number by I and compute cosh.
+        return cosh(-imaginary, real, action);
+    }
+
+    /**
+     * Returns the
+     * <a href="http://mathworld.wolfram.com/Tangent.html">
+     * tangent</a> of the complex number.
+     *
+     * <p>\[ \tan(z) = \frac{i(e^{-iz} - e^{iz})}{e^{-iz} + e^{iz}} \]
+     *
+     * <p>This is an odd function: \( \tan(z) = -\tan(-z) \).
+     * The tangent is an entire function and requires no branch cuts.
+     *
+     * <p>This is implemented using real \( x \) and imaginary \( y \) parts:</p>
+     * \[ \tan(x + iy) = \frac{\sin(2x)}{\cos(2x)+\cosh(2y)} + i \frac{\sinh(2y)}{\cos(2x)+\cosh(2y)} \]
+     *
+     * <p>As per the C99 standard this function is computed using the trigonomic identity:</p>
+     * \[ \tan(z) = -i \tanh(iz) \]
+     *
+     * @param real Real part \( a \) of the complex number \( (a +ib \).
+     * @param imaginary Imaginary part \( b \) of the complex number \( (a +ib \).
+     * @param action Consumer for the tangent of the complex number.
+     * @param <R> the return type of the supplied action.
+     * @return the object returned by the supplied action.
+     * @see <a href="http://functions.wolfram.com/ElementaryFunctions/Tan/">Tangent</a>
+     */
+    public static <R> R tan(double real, double imaginary, ComplexSink<R> action) {
+        // Define in terms of tanh
+        // tan(z) = -i tanh(iz)
+        // Multiply this number by I, compute tanh, then multiply by back
+        return tanh(-imaginary, real, (r, i) -> multiplyNegativeI(r, i, action));
+    }
+
+    /**
+     * Returns the inverse sine of the complex number.
```

Comment:
Copied the wrong javadoc. Please copy asin from Complex here.

However the attribution to Boost (and the rest of this javadoc) should not be lost. The javadoc here is from a static function previously private to Complex. So perhaps create a private one here and call it from the public `asin` function, e.g. named `computeAsin`, so all the previousl documentation on the method origin is maintained.